### PR TITLE
Undertow async result exceptions are logged with the correct traceId

### PR DESCRIPTION
**Pull request by carterkozak**
Also provides a tracing span showing asynchronous responses where
previously only the main action was traced.

## After this PR
==COMMIT_MSG==
Undertow async result exceptions are logged with the correct traceId
==COMMIT_MSG==

### CODE REVIEW

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java [hunk 1]
```diff
@@ -133,18 +134,41 @@
...
+                        serverExchange, exceptionSerializer, throwable), tracer));
             }
         }, DIRECT_EXECUTOR));
     }
+
+    private HttpHandler wrapCallback(HttpHandler action, DeferredTracer tracer) {
+        return new DeferredTracerHandler(new ConjureExceptionHandler(action, exceptionSerializer), tracer);
+    }
+
+    private static final class DeferredTracerHandler implements HttpHandler {
```
Comment (rubenfiszel) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: could also be shortened as:
```
private static final Function<DeferredTracer, HandlerWrapper> deferredTracerHandlerWrapper = tracer -> next -> tracer.withTrace(() -> {
                next.handleRequest(exchange);
                return null; })
```

**Review by rubenfiszel — APPROVED** [review]
LGTM

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java [hunk 2]
```diff
@@ -133,18 +134,41 @@
...
-                exchange.dispatch(new ConjureExceptionHandler(serverExchange -> ConjureExceptions.handle(
-                        serverExchange, exceptionSerializer, throwable), exceptionSerializer));
+                exchange.dispatch(wrapCallback(serverExchange -> ConjureExceptions.handle(
+                        serverExchange, exceptionSerializer, throwable), tracer));
             }
         }, DIRECT_EXECUTOR));
     }
+
+    private HttpHandler wrapCallback(HttpHandler action, DeferredTracer tracer) {
+        return new DeferredTracerHandler(new ConjureExceptionHandler(action, exceptionSerializer), tracer);
```
Comment (ferozco) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: Could we avoid the new class and return a lambda instead?
```suggestion
        return exchange -> tracer.withTrace(() -> {
            new ConjureExceptionHandler(action, exceptionSerializer).handleRequest(exchange);
            return null;
        });
```

Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: Updated. I find classes easier to follow in stack traces, but I don't have much preference either way.

**Comment by ferozco** [comment]
👍

**Comment by ferozco** [comment]
poke

**Pull request merged**